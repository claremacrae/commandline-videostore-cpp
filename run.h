#ifndef COMMANDLINE_VIDEOSTORE_CPP_RUN_H
#define COMMANDLINE_VIDEOSTORE_CPP_RUN_H
#include <iosfwd>
#include <string>

void run(std::istream& in, std::ostream& out);

void calculateAmountAndPoints(const std::string& category,
                              int daysRented,
                              double& thisAmount,
                              int& frequentRenterPoints);

#endif
