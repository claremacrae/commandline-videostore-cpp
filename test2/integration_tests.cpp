#include "run.h"
#include "catch2/catch.hpp"
#include "ApprovalTests.hpp"
#include <fmt/core.h>
using namespace ApprovalTests;

#include <sstream>

TEST_CASE("test empty input") {
  std::stringstream in;
  std::stringstream out;
  run(in, out);
  Approvals::verify(out.str());
}

TEST_CASE("test real input") {
  std::stringstream in;
  in << "Clare Macrae\n";
  in << "1 2\n";
  in << "1 3\n";
  std::stringstream out;
  run(in, out);
  Approvals::verify(out.str());
}

TEST_CASE("test non-production data") {
  std::vector<std::string> inputs1{"REGULAR", "CHILDRENS", "NEW_RELEASE"};
  std::vector<int> inputs2{1, 2, 3, 4};
  CombinationApprovals::verifyAllCombinations(
      "(Category, days)",
      [&](auto category, auto days) {
        double amount = 0;
        int points = 0;
        calculateAmountAndPoints(category, days, amount, points);
        return fmt::format("\t(Amount: {:7.2f}\tPoints: {}", amount, points);
      },
      inputs1,
      inputs2);
}
