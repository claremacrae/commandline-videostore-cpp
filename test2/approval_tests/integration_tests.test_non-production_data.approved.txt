(Category, days)


(REGULAR, -2) => 	(Amount:    2.00	Points: 1
(REGULAR, -1) => 	(Amount:    2.00	Points: 1
(REGULAR, 0) => 	(Amount:    2.00	Points: 1
(REGULAR, 1) => 	(Amount:    2.00	Points: 1
(REGULAR, 2) => 	(Amount:    2.00	Points: 1
(REGULAR, 3) => 	(Amount:    3.50	Points: 1
(REGULAR, 4) => 	(Amount:    5.00	Points: 1
(CHILDRENS, -2) => 	(Amount:    1.50	Points: 1
(CHILDRENS, -1) => 	(Amount:    1.50	Points: 1
(CHILDRENS, 0) => 	(Amount:    1.50	Points: 1
(CHILDRENS, 1) => 	(Amount:    1.50	Points: 1
(CHILDRENS, 2) => 	(Amount:    1.50	Points: 1
(CHILDRENS, 3) => 	(Amount:    1.50	Points: 1
(CHILDRENS, 4) => 	(Amount:    3.00	Points: 1
(NEW_RELEASE, -2) => 	(Amount:   -6.00	Points: 1
(NEW_RELEASE, -1) => 	(Amount:   -3.00	Points: 1
(NEW_RELEASE, 0) => 	(Amount:    0.00	Points: 1
(NEW_RELEASE, 1) => 	(Amount:    3.00	Points: 1
(NEW_RELEASE, 2) => 	(Amount:    6.00	Points: 2
(NEW_RELEASE, 3) => 	(Amount:    9.00	Points: 2
(NEW_RELEASE, 4) => 	(Amount:   12.00	Points: 2
(JUNK, -2) => 	(Amount:    0.00	Points: 1
(JUNK, -1) => 	(Amount:    0.00	Points: 1
(JUNK, 0) => 	(Amount:    0.00	Points: 1
(JUNK, 1) => 	(Amount:    0.00	Points: 1
(JUNK, 2) => 	(Amount:    0.00	Points: 1
(JUNK, 3) => 	(Amount:    0.00	Points: 1
(JUNK, 4) => 	(Amount:    0.00	Points: 1

